### PR TITLE
Yang la132.industry china

### PR DESCRIPTION
**Pull request by thuliuyang**

### CODE REVIEW

File: R/zchunk_LA132.Industry_china.R [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+    L101.NBS_use_all_Mtce %>%
+      filter(EBProcess == "Construction" &
+               EBMaterial == "Bitumen Asphalt") %>%
+      mutate(sector = 'industry_feedstocks', fuel = "refined liquids") ->
+      L132.in_Mtce_province_indfeed_F_unscaled.road
+
+    L101.inNBS_Mtce_province_S_F %>%
+      filter(sector == "industry_feedstocks" &
+               fuel %in% unique( L1322.in_EJ_R_indfeed_F_Yh$fuel)) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: no space after `(`

File: R/zchunk_LA132.Industry_china.R [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    # Split province-level data into feedstocks only, according to names of fuel inputs
+
+    L101.NBS_use_all_Mtce %>%
+      filter(EBProcess == "Construction" &
+               EBMaterial == "Bitumen Asphalt") %>%
+      mutate(sector = 'industry_feedstocks', fuel = "refined liquids") ->
+      L132.in_Mtce_province_indfeed_F_unscaled.road
+
+    L101.inNBS_Mtce_province_S_F %>%
+      filter(sector == "industry_feedstocks" &
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: again, replace `&` with `,`

File: R/zchunk_LA132.Industry_china.R [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      ungroup() %>%
+      replace_na(list(value = 0)) %>%
+      group_by(province, sector, fuel, year) %>%
+      summarise(value = sum(value)) %>%
+      ungroup %>%
+      replace_na(list(value = 0)) %>%
+      group_by(sector, fuel, year) %>%
+      mutate(value = value / sum(value)) %>%
+      ungroup %>%
+      replace_na(list(value = 0)) ->
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: Probably do no need three `replace_na`, please check

File: R/zchunk_LA132.Industry_china.R [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      select(-multiplier, -GCAM_region_ID) %>%
+      arrange(province) ->
+      L132.in_EJ_province_indfeed_F
+
+
+    ## OUTPUTS
+    L132.in_EJ_province_indnochp_F %>%
+      add_title("Industrial sector non-cogen input energy by province and fuel") %>%
+      add_units("EJ") %>%
+      add_comments("Computed by apportioning china-level consumption among provinces") %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: China and all below

File: R/zchunk_LA132.Industry_china.R [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      add_legacy_name("L132.in_EJ_province_indnochp_F") %>%
+      add_precursors("L1322.in_EJ_R_indenergy_F_Yh",
+                     "L101.inNBS_Mtce_province_S_F",
+                     "L122.in_EJ_province_refining_F") %>%
+      add_flags(FLAG_PROTECT_FLOAT, FLAG_SUM_TEST) ->
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: How small are the differences when you compare manually?

File: R/zchunk_LA132.Industry_china.R [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    L1322.in_EJ_R_indenergy_F_Yh <- get_data(all_data, "L1322.in_EJ_R_indenergy_F_Yh")
+    L1322.in_EJ_R_indfeed_F_Yh <- get_data(all_data, "L1322.in_EJ_R_indfeed_F_Yh")
+
+    # ===================================================
+
+    # PART 1. Compute industrial energy use, removing energy used in refining
+
+    # Aggregate all refining fuel consumption for electricity and gas by province
+    # TODO: skipping province industrial refining energy use for now due to units issues
+    if(1==2) {
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: space around "=="

File: R/zchunk_LA132.Industry_china.R [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+      L132.in_EJ_province_refining_elecgas
+
+    # Adjust industrial fuel consumption by removing refinery consumption computed above
+    L101.inNBS_Mtce_province_S_F %>%
+      filter(sector == "industry", fuel %in% c("electricity", "gas")) %>%
+      # We need to transfer the Mtce to EJ. We haven't done that yet.
+      left_join_error_no_match(L132.in_EJ_province_refining_elecgas, by = c("province", "fuel", "year")) %>%
+      #Replace any negative values with zeroes
+      mutate(value = value - refinery_comsumption,
+             value = replace(value, value <0, 0)) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: space after `<`

File: R/zchunk_LA132.Industry_china.R [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      filter(sector == "industry", fuel == "coal") %>%
+      mutate(fuel = 'biomass') ->
+      L132.in_Mtce_province_indenergy_F_unscaled.bio
+
+    # Only use the fuels that are considered in GCAM's "industrial energy use" sector
+    L101.inNBS_Mtce_province_S_F %>%
+      filter(sector == "industry",
+             fuel %in% unique(L1322.in_EJ_R_indenergy_F_Yh$fuel)) %>%
+      bind_rows(L132.in_Mtce_province_indenergy_F_unscaled.bio) %>%
+      complete(nesting(sector,fuel),province, year = HISTORICAL_YEARS) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: space after `,`

File: R/zchunk_LA132.Industry_china.R [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+      group_by(province, year) %>%
+      # First zero out NAs in years where some values are NA but not all
+      mutate(value = replace(value, is.na(value) & sum(value, na.rm = T) != 0, 0)) %>%
+      ungroup %>%
+      # use approx_fun rule = 2 to fill out data in years where the entire province is NA
+      group_by(province, sector, fuel) %>%
+      mutate( value = approx_fun(year, value, rule = 2), sum = NULL) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: no space after `(`

File: R/zchunk_LA132.Industry_china.R [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+      group_by(province, sector, fuel) %>%
+      # use approx_fun rule = 2 to fill out data in years where the entire province is NA
+      mutate(value = approx_fun(year, value, rule = 2)) %>%
+      ungroup %>%
+      # When there is one data point, approx_fun replaced it with NA, but what we needed is to replace all NAs with that data.
+      left_join(L132.in_Mtce_province_indfeed_F_unscaled %>% rename(org = value),
+                by = c("province", "fuel", "sector", "year")) %>%
+      group_by(sector, fuel, province) %>%
+      mutate(sum =sum(org, na.rm = T)) %>%
+      mutate(value = replace(value, is.na(value) & sum != 0, sum)) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: why replace with sum not the original value?

File: R/zchunk_LA132.Industry_china.R [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+      select(-org, -sum) %>%
+      ungroup %>%
+      group_by(sector, fuel, year) %>%
+      mutate(multiplier = value / sum(value, na.rm =T), value = NULL) %>%
```
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: space around `=`

File: R/zchunk_LA132.Industry_china.R (hunk 10 above)
Comment (rynacui) [inline R/zchunk_LA132.Industry_china.R]: I don't have a better solution. 
Combining the two lines this way may be slightly clearer: `mutate(value = replace(value, is.na(value) & sum(org[is.na(value), na.rm = T) != 0,  sum(org[is.na(value), na.rm = T))`
We can keep it for now and add comment explaining that only under the occasion when there is one data point in the group, approx_fun replaces the one value with NA. The original gcam_interp function replaces all missing values in that group with the one value when rule = 2.

**Comment by rynacui** [comment]
@bpbond I'm not sure why the tests on AppVeyor failed. It says the 'gcamdata.compdata' package is not available - which is strange. How to re-start the tests?

**Comment by bpbond** [comment]
@rynacui I assume that it first needs to download `gcamdata.compdata` from GitHub, not CRAN? If it's passing Travis, that's fine for now, go ahead and merge. If this persists we do need to fix it though...

**Comment by rynacui** [comment]
Okay, but we don't have this issue for all the earlier PRs, and the `DESCRIPTION` file does specify the github rep for `gcamdata.compdata`.

I'll merge this one, and let you know if we have this problem again. Thanks @bpbond !

**Pull request merged**